File: examples/python/soundside_client.py
"""
Soundside MCP Client — Python SDK

Production-grade client for Soundside's MCP endpoint with async polling,
error handling, and pipeline support.

Requirements:
    pip install httpx

Usage:
    python soundside_client.py <API_KEY>
    # or
    SOUNDSIDE_API_KEY=mcp_... python soundside_client.py
"""

import httpx
import json
import os
import sys
import time


class SoundsideClient:
    """Production MCP client for Soundside with async resource polling."""

    def __init__(self, api_key: str, endpoint: str = "https://mcp.soundside.ai/mcp"):
        self.api_key = api_key
        self.endpoint = endpoint
        self.session_id = None
        self._msg_id = 0

    def _next_id(self) -> str:
        self._msg_id += 1
        return str(self._msg_id)

    def _headers(self) -> dict:
        h = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "Accept": "application/json, text/event-stream",
        }
        if self.session_id:
            h["mcp-session-id"] = self.session_id
        return h

    def _parse_sse(self, text: str) -> dict:
        """Parse SSE response — server emits notification frames before the
        actual JSON-RPC result. Find the frame with 'id' (the response).
        Falls back to last data frame if no 'id' frame found.
        """
        last_data = None
        for line in text.splitlines():
            if line.startswith("data:"):
                try:
                    obj = json.loads(line[5:].strip())
                    # JSON-RPC responses have 'id'; notifications do not
                    if "id" in obj:
                        return obj
                    last_data = obj  # keep as fallback
                except json.JSONDecodeError:
                    pass
        if last_data is not None:
            return last_data
        # Fall back: try parsing the whole text as JSON
        return json.loads(text)

    def _extract_tool_result(self, rpc_result: dict) -> dict:
        """Extract the tool result from an MCP JSON-RPC response.

        Soundside MCP returns data in two places:
        - content[0].text — JSON string (sync tools) or plain message (async tools)
        - structuredContent — always a dict with resource_id, state, etc.

        For async tools (create_video, create_music), the resource_id is ONLY
        in structuredContent, so we merge both sources.
        """
        result = rpc_result.get("result", {})

        # Check for tool-level errors (isError=True in MCP result)
        if result.get("isError"):
            for ct in result.get("content", []):
                if ct.get("type") == "text":
                    raise RuntimeError(f"Tool error: {ct['text']}")
            raise RuntimeError("Tool error: unknown error")

        # Extract from content[0].text (may be JSON or plain text)
        text_result = {}
        content = result.get("content", [])
        for c in content:
            if c.get("type") == "text":
                try:
                    text_result = json.loads(c["text"])
                except json.JSONDecodeError:
                    text_result = {"text": c["text"]}
                break

        # Extract from structuredContent (always has resource_id for async tools)
        structured = result.get("structuredContent")
        if isinstance(structured, dict):
            # Merge: structuredContent wins for resource tracking fields
            merged = {**text_result, **structured}
            return merged

        return text_result or rpc_result

    def connect(self) -> dict:
        """Initialize MCP session."""
        with httpx.Client(timeout=30) as client:
            r = client.post(
                self.endpoint,
                json={
                    "jsonrpc": "2.0",
                    "id": self._next_id(),
                    "method": "initialize",
                    "params": {
                        "protocolVersion": "2025-11-25",
                        "capabilities": {},
                        "clientInfo": {"name": "soundside-python-sdk", "version": "1.1"},
                    },
                },
                headers=self._headers(),
            )
            self.session_id = r.headers.get("mcp-session-id")
            return self._parse_sse(r.text)

    def list_tools(self) -> list:
        """Get available tools and their schemas."""
        with httpx.Client(timeout=30) as client:
            r = client.post(
                self.endpoint,
                json={"jsonrpc": "2.0", "id": self._next_id(), "method": "tools/list", "params": {}},
                headers=self._headers(),
            )
            result = self._parse_sse(r.text)
            return result.get("result", {}).get("tools", [])

    def call_tool(self, name: str, arguments: dict, timeout: int = 120) -> dict:
        """Call an MCP tool and return the parsed result.

        Raises RuntimeError on JSON-RPC errors or tool-level errors (isError).
        """
        with httpx.Client(timeout=timeout) as client:
            r = client.post(
                self.endpoint,
                json={
                    "jsonrpc": "2.0",
                    "id": self._next_id(),
                    "method": "tools/call",
                    "params": {"name": name, "arguments": arguments},
                },
                headers=self._headers(),
            )
            rpc = self._parse_sse(r.text)
            if "error" in rpc:
                raise RuntimeError(f"MCP error: {rpc['error']}")
            return self._extract_tool_result(rpc)

    def wait_for_resource(
        self,
        resource_id: str,
        timeout: int = 300,
        poll_interval: int = 5,
    ) -> dict:
        """Poll until an async resource completes.

        Checks lib_list every `poll_interval` seconds until:
        - status="completed" AND url is present → returns the resource dict
        - status="failed" or "error" → raises RuntimeError
        - timeout exceeded → raises TimeoutError

        lib_list calls are free (zero credits). The signed GCS asset URL
        arrives on the item as ``url`` (older responses used ``storage_url``
        — both are accepted).

        Args:
            resource_id: UUID of the resource to wait for.
            timeout: Maximum seconds to wait (default 300, use 600 for video).
            poll_interval: Seconds between polls (default 5).

        Returns:
            The resource dict from lib_list (includes url, status, metadata).
        """
        start = time.time()
        while time.time() - start < timeout:
            result = self.call_tool("lib_list", {
                "entity_type": "resources",
                "resource_id": resource_id,
            })
            # lib_list wraps results: {success, status, items: [...]}
            item = result.get("items", [{}])[0] if result.get("items") else result
            status = item.get("status") or item.get("state", "")

            if status in ("failed", "error"):
                reason = item.get("failure_reason", "unknown")
                raise RuntimeError(f"Resource {resource_id} failed: {reason}")

            if status == "completed" and (item.get("url") or item.get("storage_url")):
                return item

            time.sleep(poll_interval)

        raise TimeoutError(f"Resource {resource_id} did not complete in {timeout}s")

    def call_and_wait(
        self,
        name: str,
        arguments: dict,
        timeout: int = 300,
        poll_interval: int = 5,
        call_timeout: int = 120,
    ) -> dict:
        """Call an MCP tool, then poll until the resource completes.

        Convenience wrapper for tools that return async resource_ids.
        Combines call_tool() + wait_for_resource() in one call.

        Args:
            name: Tool name (e.g., "create_video").
            arguments: Tool arguments dict.
            timeout: Max seconds to wait for resource completion.
            poll_interval: Seconds between polls.
            call_timeout: HTTP timeout for the initial tool call.

        Returns:
            The completed resource dict (includes url, status, metadata).
        """
        result = self.call_tool(name, arguments, timeout=call_timeout)
        resource_id = result.get("resource_id")
        if not resource_id:
            # Tool returned synchronously (no resource_id to poll)
            return result

        status = result.get("status") or result.get("state", "")
        asset_url = result.get("url") or result.get("storage_url")
        if status == "completed" and asset_url:
            # Already complete (sync tool or instant result)
            return result

        # Async — poll until done
        return self.wait_for_resource(resource_id, timeout=timeout, poll_interval=poll_interval)


def main():
    api_key = (
        sys.argv[1] if len(sys.argv) > 1
        else os.environ.get("SOUNDSIDE_API_KEY")
    )
    if not api_key:
        print("Usage: python soundside_client.py <API_KEY>")
        print("   or: SOUNDSIDE_API_KEY=mcp_... python soundside_client.py")
        sys.exit(1)

    client = SoundsideClient(api_key)

    # 1. Connect
    print("Connecting to Soundside MCP...")
    client.connect()
    print(f"✅ Connected (session: {client.session_id[:16]}...)")

    # 2. List tools
    tools = client.list_tools()
    print(f"\n📋 Available tools ({len(tools)}):")
    for t in tools:
        print(f"  • {t['name']}: {t.get('description', '')[:60]}")

    # 3. Generate an image (sync — returns immediately)
    print("\n🎨 Generating image (vertex)...")
    t0 = time.time()
    result = client.call_tool("create_image", {
        "prompt": "A vibrant sunset over a calm ocean, photorealistic",
        "provider": "vertex",
    })
    elapsed = time.time() - t0
    print(f"  Resource ID: {result.get('resource_id', 'N/A')}")
    print(f"  Time: {elapsed:.1f}s")
    asset_url = result.get("url") or result.get("storage_url")
    if asset_url:
        print(f"  URL: {asset_url[:80]}...")

    # 4. Analyze it
    if result.get("resource_id"):
        print("\n🔍 Analyzing image...")
        analysis = client.call_tool("analyze_media", {
            "resource_id": result["resource_id"],
        })
        meta = analysis.get("metadata", analysis)
        print(f"  Type: {meta.get('mime_type', 'N/A')}")
        if "image" in meta:
            img = meta["image"]
            print(f"  Dimensions: {img.get('width')}×{img.get('height')}")

    # 5. Generate text
    print(f"\n📝 Generating text (vertex)...")
    text_result = client.call_tool("create_text", {
        "prompt": "Write a haiku about AI agents creating art",
        "provider": "vertex",
    })
    text_output = text_result.get("message") or text_result.get("text") or str(text_result)[:200]
    print(f"  {text_output}")

    # 6. Demo: call_and_wait (generates video and polls until complete)
    print(f"\n🎬 Generating video with call_and_wait (minimax)...")
    print("   This calls create_video then automatically polls until complete.")
    if result.get("resource_id"):
        t0 = time.time()
        try:
            video = client.call_and_wait("create_video", {
                "prompt": "Gentle waves lapping at a golden shore at sunset",
                "provider": "minimax",
                "first_frame": result["resource_id"],
            }, timeout=600)
            elapsed = time.time() - t0
            print(f"  ✅ Video complete in {elapsed:.1f}s")
            print(f"  Resource ID: {video.get('id', video.get('resource_id', 'N/A'))}")
            video_url = video.get("url") or video.get("storage_url")
            if video_url:
                print(f"  URL: {video_url[:80]}...")
        except (RuntimeError, TimeoutError) as e:
            print(f"  ⚠️  Video generation failed: {e}")


if __name__ == "__main__":
    main()

File: examples/typescript/soundside-client.ts
/**
 * Soundside MCP Client — TypeScript SDK
 *
 * Production-grade client for Soundside's MCP endpoint with async polling,
 * error handling, and pipeline support.
 *
 * Requirements: npm install axios
 * Usage: npx tsx soundside-client.ts <API_KEY>
 *    or: SOUNDSIDE_API_KEY=mcp_... npx tsx soundside-client.ts
 */

import axios, { AxiosInstance, AxiosResponse } from "axios";

interface MCPResult {
  jsonrpc: string;
  id?: string;
  result?: {
    tools?: Array<{ name: string; description: string; inputSchema: object }>;
    content?: Array<{ type: string; text?: string; uri?: string; name?: string; mimeType?: string }>;
    structuredContent?: Record<string, unknown>;
    isError?: boolean;
  };
  error?: { code: number; message: string };
  method?: string; // notification frames have 'method' but no 'id'
}

interface ResourceItem {
  id?: string;
  resource_id?: string;
  status?: string;
  /** Legacy alias for status (older responses). */
  state?: string;
  /** Signed GCS URL (canonical). */
  url?: string;
  /** Legacy alias for url. */
  storage_url?: string;
  failure_reason?: string;
  [key: string]: unknown;
}

class SoundsideClient {
  private client: AxiosInstance;
  private sessionId: string | null = null;
  private msgId = 0;

  constructor(
    private apiKey: string,
    private endpoint: string = "https://mcp.soundside.ai/mcp"
  ) {
    this.client = axios.create({
      baseURL: endpoint,
      timeout: 120_000,
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json, text/event-stream",
      },
    });
  }

  private nextId(): string {
    return String(++this.msgId);
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
    };
    if (this.sessionId) {
      h["mcp-session-id"] = this.sessionId;
    }
    return h;
  }

  private parseSSE(data: string | object): MCPResult {
    if (typeof data !== "string") return data as MCPResult;

    // Server sends multiple SSE frames: notifications (with 'method') then
    // the actual JSON-RPC response (with 'id'). Find the response frame.
    let lastData: MCPResult | null = null;
    for (const line of data.split("\n")) {
      if (line.startsWith("data:")) {
        try {
          const obj: MCPResult = JSON.parse(line.slice(5).trim());
          // JSON-RPC responses have 'id'; notifications have 'method' only
          if (obj.id !== undefined) {
            return obj;
          }
          lastData = obj; // keep as fallback
        } catch {
          // skip malformed JSON
        }
      }
    }
    if (lastData !== null) {
      return lastData;
    }
    return JSON.parse(data);
  }

  private extractToolResult(rpc: MCPResult): Record<string, unknown> {
    if (rpc.error) return { error: rpc.error };

    // Check for tool-level errors (isError=true in MCP result)
    if (rpc.result?.isError) {
      const content = rpc.result?.content ?? [];
      for (const c of content) {
        if (c.type === "text" && c.text) {
          throw new Error(`Tool error: ${c.text}`);
        }
      }
      throw new Error("Tool error: unknown error");
    }

    // Prefer structuredContent (MCP 2025-11-25 format)
    if (rpc.result?.structuredContent) {
      const sc = rpc.result.structuredContent;
      // FastMCP wraps returns in a 'result' key — unwrap if needed
      if (sc.result !== undefined && Object.keys(sc).length === 1) {
        return sc.result as Record<string, unknown>;
      }
      return sc;
    }

    // Fall back to content blocks
    const content = rpc.result?.content ?? [];
    let parsed: Record<string, unknown> = {};

    // Extract text content
    for (const c of content) {
      if (c.type === "text" && c.text) {
        try {
          parsed = JSON.parse(c.text);
        } catch {
          parsed = { text: c.text };
        }
        break;
      }
    }

    // Extract resource_link content blocks (MCP 2025-11-25)
    const resourceLinks = content
      .filter((c) => c.type === "resource_link")
      .map((c) => ({ uri: c.uri, name: c.name, mimeType: c.mimeType }));
    if (resourceLinks.length > 0) {
      parsed._resourceLinks = resourceLinks;
    }

    return Object.keys(parsed).length > 0
      ? parsed
      : (rpc as unknown as Record<string, unknown>);
  }

  async connect(): Promise<void> {
    const { data, headers }: AxiosResponse = await this.client.post(
      "",
      {
        jsonrpc: "2.0",
        id: this.nextId(),
        method: "initialize",
        params: {
          protocolVersion: "2025-11-25",
          capabilities: {},
          clientInfo: { name: "soundside-ts-sdk", version: "1.1" },
        },
      },
      { headers: this.headers() }
    );
    this.sessionId = headers["mcp-session-id"] ?? null;
    console.log(
      `✅ Connected (session: ${this.sessionId?.slice(0, 16)}...)`
    );
  }

  async listTools(): Promise<
    Array<{ name: string; description: string }>
  > {
    const { data } = await this.client.post(
      "",
      {
        jsonrpc: "2.0",
        id: this.nextId(),
        method: "tools/list",
        params: {},
      },
      { headers: this.headers() }
    );
    const result = this.parseSSE(data);
    return result.result?.tools ?? [];
  }

  async callTool(
    name: string,
    args: Record<string, unknown>,
    timeout?: number
  ): Promise<Record<string, unknown>> {
    const { data } = await this.client.post(
      "",
      {
        jsonrpc: "2.0",
        id: this.nextId(),
        method: "tools/call",
        params: { name, arguments: args },
      },
      {
        headers: this.headers(),
        timeout: timeout ?? 120_000,
      }
    );
    const result = this.parseSSE(data);
    return this.extractToolResult(result);
  }

  /**
   * Poll until an async resource completes.
   *
   * Checks lib_list every `pollIntervalMs` until:
   * - status="completed" AND url is present → returns the resource
   * - status="failed" or "error" → throws Error
   * - timeout exceeded → throws Error
   *
   * lib_list calls are free (zero credits). The signed GCS asset URL
   * arrives on the item as ``url``; older responses used ``storage_url``.
   * Both are accepted.
   */
  async waitForResource(
    resourceId: string,
    timeoutMs: number = 300_000,
    pollIntervalMs: number = 5_000
  ): Promise<ResourceItem> {
    const start = Date.now();
    while (Date.now() - start < timeoutMs) {
      const result = await this.callTool("lib_list", {
        entity_type: "resources",
        resource_id: resourceId,
      });

      const items = (result.items ?? []) as ResourceItem[];
      const item: ResourceItem = items[0] ?? (result as ResourceItem);
      const status = item.status ?? item.state ?? "";

      if (status === "failed" || status === "error") {
        throw new Error(
          `Resource ${resourceId} failed: ${item.failure_reason ?? "unknown"}`
        );
      }

      if (status === "completed" && (item.url || item.storage_url)) {
        return item;
      }

      await new Promise((r) => setTimeout(r, pollIntervalMs));
    }

    throw new Error(
      `Resource ${resourceId} did not complete in ${timeoutMs / 1000}s`
    );
  }

  /**
   * Call a tool, then poll until the resource completes.
   *
   * Convenience wrapper for async tools (create_video, create_music, etc.)
   * that combines callTool() + waitForResource().
   */
  async callToolAndWait(
    name: string,
    args: Record<string, unknown>,
    timeoutMs: number = 300_000,
    pollIntervalMs: number = 5_000,
    callTimeoutMs?: number
  ): Promise<Record<string, unknown>> {
    const result = await this.callTool(name, args, callTimeoutMs);
    const resourceId = result.resource_id as string | undefined;

    if (!resourceId) {
      // Sync tool — no resource_id to poll
      return result;
    }

    const status = (result.status as string) ?? (result.state as string) ?? "";
    const assetUrl = (result.url as string) ?? (result.storage_url as string);
    if (status === "completed" && assetUrl) {
      // Already complete
      return result;
    }

    // Async — poll until done
    return await this.waitForResource(resourceId, timeoutMs, pollIntervalMs);
  }
}

// --- Main ---

async function main() {
  const apiKey =
    process.argv[2] ?? process.env.SOUNDSIDE_API_KEY;
  if (!apiKey) {
    console.error("Usage: npx tsx soundside-client.ts <API_KEY>");
    console.error(
      "   or: SOUNDSIDE_API_KEY=mcp_... npx tsx soundside-client.ts"
    );
    process.exit(1);
  }

  const client = new SoundsideClient(apiKey);

  // 1. Connect
  await client.connect();

  // 2. List tools
  const tools = await client.listTools();
  console.log(`\n📋 Available tools (${tools.length}):`);
  for (const t of tools) {
    console.log(`  • ${t.name}: ${(t.description ?? "").slice(0, 60)}`);
  }

  // 3. Generate an image (sync — returns immediately)
  console.log("\n🎨 Generating image (vertex)...");
  const t0 = Date.now();
  const imageResult = await client.callTool("create_image", {
    prompt: "A vibrant sunset over a calm ocean, photorealistic",
    provider: "vertex",
  });
  const elapsed = ((Date.now() - t0) / 1000).toFixed(1);
  console.log(`  Resource ID: ${imageResult.resource_id ?? "N/A"}`);
  console.log(`  Time: ${elapsed}s`);
  const imageUrl = (imageResult.url ?? imageResult.storage_url) as string | undefined;
  if (imageUrl) {
    console.log(`  URL: ${imageUrl.slice(0, 80)}...`);
  }

  // 4. Analyze it
  if (imageResult.resource_id) {
    console.log("\n🔍 Analyzing image...");
    const analysis = await client.callTool("analyze_media", {
      resource_id: imageResult.resource_id,
    });
    const meta = (analysis.metadata ?? analysis) as Record<string, unknown>;
    console.log(`  Type: ${meta.mime_type ?? "N/A"}`);
    const img = meta.image as Record<string, unknown> | undefined;
    if (img) {
      console.log(`  Dimensions: ${img.width}×${img.height}`);
    }
  }

  // 5. Generate text
  console.log("\n📝 Generating text (vertex)...");
  const textResult = await client.callTool("create_text", {
    prompt: "Write a haiku about AI agents creating art",
    provider: "vertex",
  });
  console.log(
    `  ${String(textResult.text ?? JSON.stringify(textResult)).slice(0, 200)}`
  );

  // 6. Demo: callToolAndWait (generates video and polls until complete)
  if (imageResult.resource_id) {
    console.log("\n🎬 Generating video with callToolAndWait (minimax)...");
    console.log("   This calls create_video then automatically polls until complete.");
    const vt0 = Date.now();
    try {
      const video = await client.callToolAndWait(
        "create_video",
        {
          prompt: "Gentle waves lapping at a golden shore at sunset",
          provider: "minimax",
          first_frame: imageResult.resource_id,
        },
        600_000 // 10 min timeout for video
      );
      const velapsed = ((Date.now() - vt0) / 1000).toFixed(1);
      console.log(`  ✅ Video complete in ${velapsed}s`);
      console.log(`  Resource ID: ${video.id ?? video.resource_id ?? "N/A"}`);
      const videoUrl = (video.url ?? video.storage_url) as string | undefined;
      if (videoUrl) {
        console.log(`  URL: ${videoUrl.slice(0, 80)}...`);
      }
    } catch (err) {
      console.log(`  ⚠️  Video generation failed: ${err}`);
    }
  }
}

main().catch(console.error);
